internal/crypto/crypto.go

```diff
@@ -69,3 +73,137 @@ func GenerateSignatures(secrets []string, msgID uuid.UUID, currentTime time.Time
 	}
 	return signatureList, nil
 }
+
+type EncryptedString struct {
+	KeyID     string `json:"key_id"`
+	Algorithm string `json:"alg"`
+	Data      []byte `json:"data"`
+	Nonce     []byte `json:"nonce,omitempty"`
+}
+
+func (es *EncryptedString) IsValid() bool {
+	return es.KeyID != "" && len(es.Data) > 0 && len(es.Nonce) > 0 && es.Algorithm == "aes-gcm-hkdf"
+}
+
+// ShouldReEncrypt tells you if the value encrypted needs to be encrypted again with a newer key.
+func (es *EncryptedString) ShouldReEncrypt(encryptionKeyID string) bool {
+	return es.KeyID != encryptionKeyID
+}
+
+func (es *EncryptedString) Decrypt(id string, decryptionKeys map[string]string) ([]byte, error) {
+	decryptionKey := decryptionKeys[es.KeyID]
+
+	if decryptionKey == "" {
+		return nil, fmt.Errorf("crypto: decryption key with name %q does not exist", es.KeyID)
+	}
+
+	key, err := deriveSymmetricKey(id, es.KeyID, decryptionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	aes, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	cipher, err := cipher.NewGCM(aes)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypted, err := cipher.Open(nil, es.Nonce, es.Data, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return decrypted, nil
+}
+
+func ParseEncryptedString(str string) *EncryptedString {
+	if !strings.HasPrefix(str, "{") {
+		return nil
+	}
+
+	var es EncryptedString
+
+	if err := json.Unmarshal([]byte(str), &es); err != nil {
+		return nil
+	}
+
+	if !es.IsValid() {
+		return nil
+	}
+
+	return &es
+}
+
+func (es *EncryptedString) String() string {
+	out, err := json.Marshal(es)
+	if err != nil {
+		panic(err)
+	}
+
+	return string(out)
+}
+
+func deriveSymmetricKey(id, keyID, keyBase64URL string) ([]byte, error) {
+	hkdfKey, err := base64.RawURLEncoding.DecodeString(keyBase64URL)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hkdfKey) != 256/8 {
+		return nil, fmt.Errorf("crypto: key with ID %q is not 256 bits", keyID)
+	}
+
+	// Since we use AES-GCM here, the same symmetric key *must not be used
+	// more than* 2^32 times. But, that's not that much. Suppose a system
+	// with 100 million users, then a user can only change their password
```

Comment:
oh okay looks like is because 100 million ~2^26 and  42 ~ 2^5.5  and 

2^26 * 2^5.5 ~= 2^32